[Spring 지하철 노선도 - 3단계] 디우(김동규) 미션 제출합니다.

tco0427 opened the pull request:
안녕하세요! 코니!! 디우입니다. 😃

이번 3단계 미션의 요구사항이 어려워서..생각보다 PR이 늦어졌네요ㅠ.ㅠ🥲
많은 부분에서 부족하지만, 코니와 조금이라도 더 많이 이야기하고 싶어 PR 제출합니다..!!

---

3단계에서 Section(구간) 들을 관리하는 Sections를 두고, 해당 도메인에서 비즈니스 로직을 처리하게 하였습니다. 그러다 보니 DB에 추가하거나 할 때, Sections를 활용하여야 했고 결국 LineId에 해당하는 section들을 모두 지운 후 새롭게 저장하는 방법 밖에 떠오르지 않더라구요..그런데 각 데이터에 대해서 쿼리를 날리는 것이 부담스러워 `batchUpdate`를 활용하였습니다.

또한 1,2 단계에서 피드백 주셨던 부분(https://github.com/woowacourse/atdd-subway-map/pull/208#discussion_r867317285) 에 대해서도 고민을 해보았는데...이전에 말씀 드렸던 것 처럼 DAO에서 다른 커스텀 예외나 표준예외를 던지는 것 이외에는 좋은 방법이 떠오르지 않더라구요...ㅠㅠ 조금만 더 힌트를 주시면 감사하겠습니다..!(EmptyResultDataAccessException에 대해서는 DAO에서 예외를 잡고, Optional을 활용해보았습니다!)

---

이번 3단계도 잘부탁드립니다!! 감사합니다.😃

tco0427 committed:
feat: 인수테스트 Fixture 적용

RestAssured를 활용하여 REST API를 테스트하는 부분을 Fixture로 분리했습니다.

tco0427 committed:
feat: id 추출부분을 별도의 메소드로 분리

id를 추출하는 부분을 별도의 메소드로 분리하여 코드의 가독성을 개선해보았습니다.

tco0427 committed:
refactor: Dao 인터페이스 도입

DAO 의 기능명세를 위한 인터페이스를 도입하였습니다.

tco0427 committed:
feat: Fake DAO 적용

가짜 DAO객체를 통해서 Service를 테스트하도록 하였습니다.

tco0427 committed:
refactor: 사용하지 않는 생성자 제거

사용하지 않는 기본생성자를 제거하였습니다.

tco0427 committed:
refactor: 생성자 추가

Line 도메인을 받아와 초기화를 진행하는 생성자를 추가하였습니다.

tco0427 committed:
feat: 인자 크기 검증 추가

지정된 컬럼의 크기를 초과하는 값에 대한 처리를 추가해주었습니다.

tco0427 committed:
docs(README): 지하철 구간 관리 기능 요구사항 정리

tco0427 committed:
feat: Section 도메인

Section의 두 종점 간의 거리는 0보다 작거나 같을 수 없으며, 상행종점과 하행 종점은 같은 역일 수 없다.

tco0427 committed:
feat: 구간 등록

새로 등록할 구간의 상행역과 하행역 중 노선에 이미 등록되어 있는 역을 기준으로 새로운 구간을 추가한다.

tco0427 committed:
feat: 구간 등록

하나의 노선에는 갈래길이 허용되지 않기 때문에 새로운 구간이 추가되기 전에 갈래길이 생기지 않도록 기존 구간을 변경한다.

tco0427 committed:
feat: 구간 등록

역 사이에 새로운 역을 등록할 경우 기존 역 사이 길이보다 크거나 같으면 등록을 할 수 없다.

tco0427 committed:
feat: 구간 등록

상행역과 하행역이 이미 노선에 모두 등록되어 있다면 추가할 수 없다.

tco0427 committed:
feat: 구간 등록

상행역과 하행역 둘 중 하나도 포함되어 있지 않으면 추가할 수 없다.

tco0427 committed:
feat: 구간 제거

Station을 통해서 구간을 제거할 수 있다.

tco0427 committed:
feat: 구간 제거

구간이 하나인 노선에서 마지막 구간을 제거할 수 없다.

tco0427 committed:
feat: SectionDao 구현

tco0427 committed:
feat: Line에 Sections 추가 및 생성자 추가

tco0427 committed:
feat: 정렬된 Section 리스트 반환

tco0427 committed:
feat: response DTO에 Station에 대한 내용 추가

tco0427 committed:
feat: LineService 기존 테스트 통과하도록 수정

tco0427 committed:
feat: E2E Line 테스트시 변경된 사항(station정보) 추가하도록 수정

tco0427 committed:
feat: 단건 노선에 대한 구간 조회 구현

tco0427 committed:
feat: SectionService 등록 및 삭제 구현

tco0427 committed:
feat: SectionController 구현

tco0427 committed:
feat: 불필요한 코드 제거 및 구간 추가시 검사

tco0427 committed:
test: 환승역(교점)에 대한 구간 테스트

tco0427 committed:
feat: Sections를 이용하도록 수정

tco0427 committed:
feat: 구간 등록 테스트 추가

tco0427 committed:
refactor: 불필요한 생성자 및 필드 제거

tco0427 committed:
refactor: 불필요한 생성자 및 필드 제거

tco0427 committed:
refactor: EmptyResultDataAccessException 제거

tco0427 committed:
fix: 노선 등록시 최초의 구간을 등록하므로 이를 검사하는 로직 제거

tco0427 committed:
feat: batchUpdate 적용

tco0427 committed:
fix: 갈래길 방지 로직 수정

하행이 같은 경우에 대해서 처리를 못했었던 문제를 수정하였습니다.

tco0427 committed:
feat: 노선 삭제시 구간들 함께 삭제

tco0427 committed:
feat: 역 이전에 구간 먼저 삭제

역을 삭제하기 위해서는 해당 역을 사용중인 구간을 먼저 삭제해야만 삭제 가능하도록 구현하였습니다.

tco0427 committed:
feat: 처음과 끝 역 삭제 구현

처음과 끝 역에 대해서도 삭제가 가능하도록 로직을 추가해주었습니다.

tco0427 committed:
refactor: PathVariable 생략

choihz commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -24,16 +24,13 @@ class LineAcceptanceTest extends AcceptanceTest {
     void createLine() {
         // given
-        final LineRequest params = new LineRequest("신분당선", "bg-red-600");
+        final Long upStationId = extractStationIdFromName("지하철역");
+        final Long downStationId = extractStationIdFromName("새로운지하철역");
```
처음에 메서드 이름만 보고서는 이 코드가 하는 역할을 이해하기 조금 어려웠는데요, 해당 메서드가 하는 본질적인 일이 `extract`일까요?

tco0427 replied:
1. 지하철역 이름(name)을 통해서 `StationRequest` 를 만들고, 요청을 보낸 이후 `ExtractableResponse` 를 만든다.
2. 해당 응답에서 id값을 추출한다.

이렇게 2가지 책임을 가지기 때문에 해당 메소드가 하는 일을 `메소드 이름`을 통해서 얻기가 어려웠다고 생각이 듭니다.
두 책임은 엄연히 다른 책임이고, `makeStationResponse` 와 `extractStationId` 두 가지 메소드로 분리하여 `메소드 이름`을 통해서 그 의미를 보다 명확히 알 수 있도록 개선해보았습니다!😃

choihz commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -74,25 +60,15 @@ void createLineWithDuplicateName() {
+        ExtractableResponse<Response> createResponse2 = AcceptanceFixture.post(params2, "/lines");
 
         // when
         ExtractableResponse<Response> response = RestAssured.given().log().all()
```
여긴 변경이 누락된 걸까요?

tco0427 replied:
앗,, `AcceptanceFixture.get()` 부분을 말씀해주신게 맞을까요?? 미처 캐치하지 못했었네요..!!ㅠ.ㅠ
수정하도록 하겠습니다!! 감사합니다😃

choihz commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -103,37 +79,26 @@ void getLines() {
     }
 
     @Test
     @DisplayName("노선을 조회한다.")
```
무엇으로 노선을 조회한다는 것인지 조금 더 분명하게 표현해 준다면 좋을 것 같아요.

tco0427 replied:
`@DisplayName` 부분을 말씀해주신게 맞으실까요??
"노선 id값을 통해서 노선을 조회한다." 라고 수정해보았습니다😁

choihz commented on src/test/java/wooteco/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,116 @@
+
+public class SectionAcceptanceTest extends AcceptanceTest {
+
+    @DisplayName("환승역(교점)에 대한 구간 테스트")
```
이 이름만 봐서는 무엇을 테스트하고 있는지 코드까지 다 봐야 알 수 있을 것 같아요. 작성할 때 조금만 더 신경 쓰면 나중에 정말 많은 시간을 아낄 수 있어요!

tco0427 replied:
구현에 바쁘다보니...전체적으로 네이밍등에 신경을 못 쓴 것 같습니다..ㅠ.ㅠ
`작성할 때 조금만 더 신경 쓰면 나중에 정말 많은 시간을 아낄 수 있어요!` 말과 같이 가독성 있는 코드에 대해서는 정말 많이 공감합니다!
최초에 작성할 때에도 어느정도 신경을 써야하는데, 아직까지 이런 습관이 몸에 베지는 못한 것 같아요🥲
(테스트코드는 하나의 문서 역할을 하기 때문에 더욱 신경을 써야하는데, 특히 잘 못지키는 것 같아요 ㅠ.ㅠ)
앞으로 더욱 신경쓸 수 있도록 하겠습니다..!! 이러한 부분까지 꼼꼼히 봐주셔서 감사합니다🙏

choihz commented on src/test/java/wooteco/subway/dao/jdbc/JdbcLineDaoTest.java:
```diff
@@ -61,10 +62,11 @@ void findAll() {
         // when
-        final Line findLine = lineDao.findById(savedLine.getId());
+        final Line findLine = jdbcLineDao.findById(savedLine.getId())
+                .orElseThrow(() -> new NotExistException("찾으려는 간선이 존재하지 않습니다."));
```
`orElseThrow()` 부분이 이 테스트 코드에서 꼭 필요한가요?

tco0427 replied:
말씀해주신대로 굳이 예외를 던질 필요는 없어보입니다. 예외를 던지기 보다는 "실패"를 의미할 수 있는 객체를 만들어서 orElseGet()을 이용해 null인 경우에는 해당 객체를 반환하도록 함으로써 테스트를 해볼 수 있을 것 같습니다..!! 😃 수정해보도록하겠습니다.

choihz commented on src/main/java/wooteco/subway/domain/Station.java:
```diff
@@ -24,6 +21,9 @@ private void validateArgument(String name) {
         if (name.isBlank()) {
             throw new IllegalArgumentException("지하철역 이름이 공백일 수 없습니다.");
         }
+        if (name.length() >= 255) {
+            throw new IllegalArgumentException("지하철역 이름이 너무 깁니다.");
+        }
```
이런 상황에서는 "너무 길다"는 추상적인 표현을 사용하기보다, 정확히 몇 자까지 허용되는지 알려준다면 훨씬 좋지 않을까요? 또한 테스트 코드도 추가해 보아요.

tco0427 replied:
넵!! 동의합니다! 테스트 코드 또한 추가해보았으며, `StationName` 이라는 VO를 만들어 해당 VO에서 이름에 대한 검증을 하도록 하였습니다.😁

choihz commented on src/main/java/wooteco/subway/domain/Section.java:
```diff
@@ -0,0 +1,76 @@
+        checkSameStation(upStationId, downStationId);
+    }
+
+    private void checkNegativeDistance(int distance) {
+        if (distance <= 0) {
+            throw new IllegalSectionException("두 종점간의 거리는 0보다 작거나 같을 수 없습니다.");
+        }
+    }
```
우리의 비즈니스 로직에서 "종점"이란 용어는 노선의 양 끝 역을 나타내는 것으로 아는데, 구간은 종점만 연결하는 개념이 아니지 않나요?

tco0427 replied:
맞습니다...😅 용어를 제대로 사용하지 못했네요...수정하였습니다!

choihz commented on src/main/java/wooteco/subway/domain/Section.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    public boolean isContainStation(Section section) {
+        return upStationId.equals(section.getDownStationId()) || downStationId.equals(section.getUpStationId());
+    }
```
[Bool 변수 이름 제대로 짓기 위한 최소한의 영어 문법](https://soojin.ro/blog/naming-boolean-variables)을 한번 읽어 보면 좋을 것 같아요!

> ❗주의❗is로 시작하는 변수명을 짓다가 범하는 흔한 실수가 바로 is + 동사원형 을 쓰는 것이다.
> 
> isAuthorize, isHide, isFind 등등.

tco0427 replied:
해당 케스트는 "is + 동사원형" 을 사용한 경우입니다. 따라서 링크걸어주신 블로그 글의 마지막 내용인 "동사원형 용법" 을 참고하여 `containsStation` 으로 네이밍을 수정하여 "역을 포함하는가?" 라는 의미를 전달할 수 있도록 수정해보았습니다!

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,217 @@
+
+public class Sections {
+
+    private List<Section> sections;
```
```suggestion
    private final List<Section> sections;
```

tco0427 replied:
수정하도록 하겠습니다!!😅

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,217 @@
+        sections.remove(existingSection);
+    }
+
+    private void sameUpstationInFork(Section existingSection, Section newSection) {
```
```suggestion
    private void sameUpStationInFork(Section existingSection, Section newSection) {
```

tco0427 replied:
헉,,,수정하도록 하겠습니다!!

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,217 @@
+        }
+    }
+
+    private void checkDistance(Section existingSection, Section newSection) {
+        if (existingSection.getDistance() <= newSection.getDistance()) {
+            throw new IllegalSectionException("구간 등록이 불가능합니다.");
+        }
+    }
+
+    private void isPossibleRegistration(Section section) {
+        sections.stream()
+                .filter(s -> s.isContainStation(section))
+                .findAny()
+                .orElseThrow(() -> new IllegalSectionException("구간 등록이 불가능합니다."));
+    }
```
다른 예외 상황에서 같은 타입의 예외, 같은 메시지를 사용한다면 예외 발생 시 원인을 파악할 수 있을까요?

tco0427 replied:
`IllegalSectionException` 은 구간(section)과 관련된 예외를 나타내도록 커스텀 예외를 만들었습니다. 또한 두 경우 모두 "구간 등록이 불가능한 상황"이라는 동일한 예외 상황이라고 생각이 듭니다.(예외의 발생 이유(길이 제한, 등록 조건)는 서로 다르지만, 동일하게 사용자의 잘못된 입력 요청, bad request)
따라서 예외 타입은 동일한 예외 타입을 사용하는 대신, "어떤 이유로" 구간 등록이 불가능한지를 명확히 하는 메시지로 변경해주었습니다..!

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,217 @@
+        if (existingSection.isSameSection(section)) {
+            throw new IllegalSectionException("이미 구간이 등록되어 있습니다.");
+        }
+    }
```
메서드의 이름이 실제 하는 일을 잘 드러내고 있나요?

tco0427 replied:
스트림을 적용하여 불필요한 메소드를 제거하고 코드를 간략화하였습니다.😁

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,217 @@
+    private void processFork(Section existingSection, Section newSection) {
+        checkDistance(existingSection, newSection);
+
+        sameUpstationInFork(existingSection, newSection);
+        sameDownStationInFork(existingSection, newSection);
```
이 메서드들은 무슨 일을 **하는지** 이름으로 알 수가 없네요.

tco0427 replied:
이전처럼 "상행역이 같은 경우에 대해 처리", "하행역이 같은 경우에 대해 처리"로 분리하는 것이 아니라 "중간에 새로운 구간 추가", "새로운 구간 만들기" 라는 두 가지 로직을 메소드 분리를 통해 분리해냄으로써 의미를 보다 명확히 하도록 수정해보았습니다.

```java
    private void addSectionInMiddle(Section existingSection, Section newSection) {
        Section section = createNewSection(existingSection, newSection);

        sections.add(section);
    }

    private Section createNewSection(Section existingSection, Section newSection) {
        if (existingSection.getUpStationId().equals(newSection.getUpStationId())) {
            return new Section(existingSection.getId(), existingSection.getLineId(), newSection.getDownStationId(),
                    existingSection.getDownStationId(), existingSection.getDistance() - newSection.getDistance());
        }

        return new Section(existingSection.getId(), existingSection.getLineId(), existingSection.getUpStationId(),
                newSection.getUpStationId(), existingSection.getDistance() - newSection.getDistance());
    }
```

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,217 @@
+        return sections;
+    }
+
+    private void validateDeletion() {
+        if (sections.size() <= 1) {
+            throw new IllegalSectionException("구간을 제거할 수 없습니다.");
+        }
+    }
```
여기서도, 메서드의 이름과 예외 메시지에 상황을 조금 더 분명하게 드러내 보아요.

tco0427 replied:
고민을 해보았는데요...`validateMinimumSectionCount`, `validateThatHasMinimumSection` 를 후보(?)로 했습니다...의미만 보면 `validateThatHasMinimumSection` 이 나은 것 같은데 메소드명이 너무 길어도 안좋지 않나??하는 생각이 들었어요. 코니는 어떻게 생각하시나요??
(hasMinimumSection도 괜찮은 것 같다는 생각도 듭니다.🙂)

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,217 @@
+        final Section previousSection = getPreviousSection(stationId);
+        final Section laterSection = getLaterSection(stationId);
+
+        findAndRemoveFirstSection(stationId, previousSection);
+
+        findAndRemoveLastSection(stationId, laterSection);
+
+        removeMiddleSection(previousSection, laterSection);
```
```suggestion
        findAndRemoveFirstSection(stationId, previousSection);
        findAndRemoveLastSection(stationId, laterSection);
        removeMiddleSection(previousSection, laterSection);
```

공백은 불필요해 보여요.

tco0427 replied:
저는 보통 각 개념이 분리된다 라고 생각될 때 공백으로 구분해주려고 하는데요!, "Section을 제거하는 로직" 이라는 큰 틀에서 개념적으로 공통된 것 같아 코니의 의견대로 공백을 제거해도 될 것 같습니다!! 수정해보도록 하겠습니다!😄

choihz commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -16,32 +23,79 @@ public class LineService {
+    }
+
+    private List<StationResponse> makeStationResponseList(LineRequest request) {
+        final Station upStation = stationDao.findById(request.getUpStationId())
+                .orElseThrow(() -> new NotExistException("찾으려는 역이 존재하지 않습니다."));
+        final Station downStation = stationDao.findById(request.getDownStationId())
+                .orElseThrow(() -> new NotExistException("찾으려는 역이 존재하지 않습니다."));
```
불필요한 중복이 너무 많지 않나요?

tco0427 replied:
메소드 추출을 통해서 중복되는 부분을 제거해보았습니다!😁

choihz commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -16,32 +23,79 @@ public class LineService {
 
     private final LineDao lineDao;
+    private final SectionDao sectionDao;
+    private final StationDao stationDao;
 
-    public LineService(LineDao lineDao) {
+    public LineService(LineDao lineDao, SectionDao sectionDao, StationDao stationDao) {
         this.lineDao = lineDao;
+        this.sectionDao = sectionDao;
+        this.stationDao = stationDao;
     }
 
     public LineResponse save(LineRequest request) {
         Line line = new Line(request.getName(), request.getColor());
-
         final Line savedLine = lineDao.save(line);
 
-        return new LineResponse(savedLine.getId(), savedLine.getName(), savedLine.getColor());
+        final Section section = new Section(savedLine.getId(), request.getUpStationId(), request.getDownStationId(),
+                request.getDistance());
+        sectionDao.save(section);
+
+        return new LineResponse(savedLine, makeStationResponseList(request));
+    }
```
`Line`과 `Section` 모두 무사히 저장되어야 이 메서드의 임무가 완수되었다고 볼 수 있을 것 같고, 만약 `Section`을 저장하는 데 실패했다면 `Line`의 저장도 롤백되어야 하지 않을까요? 지금 `LineRequest`에 `name`, `color`만 채워서 요청을 보내 보고 어떤 일이 일어나고 있는지 확인해 볼까요?

tco0427 replied:
선언적 트랜잭션 처리를 지원해주는 어노테이션인 `@Transactional` 을 이용하여 해결해주었습니다.
트랜잭션은 하나의 논리적인 작업 단위로 말씀해주신 것과 같이 Line 과 Section이 모두 저장 되어야 하나의 논리적인 작업(위의 메소드)이 수행되었다고 이야기할 수 있습니다.

Transactional 어노테이션을 이용하면 데이터 추가, 갱신, 삭제 등 DB 컬럼에 변경이 이루어지는 작업 도중 오류가 발생했을 때 모든 작업들을 원상태로 되돌릴 수 있습니다. (@Transactional이 붙은 메소드는 메소드가 포함하고 있는 작업 중에 하나라도 실패할 경우 전체 작업을 취소한다.) 이를 Service 클래스에 붙여주고, 읽기만 하는 경우(ex. findById, findAll)에는 readOnly = true 를 주어 해당 트랜잭션을 읽기 전용으로 설정해주도록 하였습니다.
(@Transactional 은 클래스, 메소드 모두에 걸어줄 수 있고 메소드 레벨의 @Transactional 선언이 우선 적용된다.)

다음의 블로그 글을 참고하여 공부하였습니다. [Transactional 어노테이션](https://tecoble.techcourse.co.kr/post/2021-05-25-transactional/)

choihz reviewed (changes requested):
안녕하세요, 디우!

지하철 노선도 미션의 3단계도 잘 구현해 주셨네요 👍
다만 내부 로직까지 꼼꼼하게 보려니 리뷰가 너무 오래 걸릴 것 같아서, 우선 눈에 보이는 부분부터 코멘트 남겼어요. 내부 로직에 대한 조금 더 디테일한 리뷰와 주신 질문에 대한 답은 다시 남기도록 할게요.

궁금한 점이나 의견이 있으시다면 언제든 코멘트 또는 DM 주세요~

tco0427 committed:
refactor: 의미를 명확히 하기 위해 메소드 분리

`extractStationIdFromName` 이라는 메소드는 "ExtractableResponse 만들기" 와 "id값 추출" 이라는 두가지 책임을 가지고 있었으므로 이를 `makeStationResponse` 와 `extractStationId` 두 개의 메소드로 분리하여 그 의미를 명확히 하였습니다.

tco0427 committed:
refactor: AcceptanceFixture 사용

기존에 AcceptanceFixture 재사용이 누락된 부분에 AcceptanceFixture.get() 을 적용하였습니다.

tco0427 committed:
refactor: DisplayName 을 구체적으로 수정

DisplayName을 수정하여 무엇을 테스트하는 것인지를 보다 명확히 알 수 있도록 하였습니다.

tco0427 committed:
refactor: 테스트 코드에서의 orElseThrow 제거

orElseThrow를 제거하고 대신 "실패"를 의미하는 객체를 생성하여 orElseGet()으로 null인경우에만 반환하도록 하였습니다.

tco0427 committed:
feat: 지하철역 이름 검사에 대한 테스트 추가

지하철역이 255를 초과하는 경우에 대한 테스트 코드 추가

tco0427 committed:
feat: StationName VO 적용

tco0427 committed:
refactor: 예외 메시지 변경

종점은 노선의 양 끝역을 이야기하므로, 제대로된 예외 메시지가 아니어서 이를 수정해주었습니다.

tco0427 committed:
refactor: 메소드명 변경

isContainStation은 `is + 동사원형` 형태이므로 `containsStation` 으로 수정하였습니다.

tco0427 committed:
refactor: final 키워드 추가

tco0427 committed:
refactor: 불필요한 공백 제거

tco0427 committed:
refactor: 메소드 분리 및 메소드명 변경

tco0427 committed:
refactor: 중복되는 부분에 대해서 메소드 추출

tco0427 committed:
refactor: 직관적인 메소드명으로 변경

tco0427 committed:
refactor: 예외 메시지를 구체적으로 수정

tco0427 committed:
refactor: 코드 간략화

스트림을 적용하여 메소드를 제거하고, 변수명 수정

tco0427 committed:
refactor: @Transactional 어노테이션 추가

@Transactional 어노테이션을 추가하여 하나의 메소드가 하나의 트랙젼션이 되도록 하였습니다.

tco0427 committed:
refactor: 메소드 순서 정리

choihz commented on src/test/java/wooteco/subway/dao/jdbc/JdbcLineDaoTest.java:
```diff
@@ -61,10 +63,11 @@ void findAll() {
 
         // when
-        final Line findLine = lineDao.findById(savedLine.getId());
+        final Line findLine = jdbcLineDao.findById(savedLine.getId())
+                .orElseGet(() -> new Line(FAIL_FIND_LINE, FAIL_FIND_LINE));
```
> 말씀해주신대로 굳이 예외를 던질 필요는 없어보입니다. 예외를 던지기 보다는 "실패"를 의미할 수 있는 객체를 만들어서 orElseGet()을 이용해 null인 경우에는 해당 객체를 반환하도록 함으로써 테스트를 해볼 수 있을 것 같습니다..!! 😃 수정해보도록하겠습니다.

여기서 `Optional`이 비었을 때 예외를 던지는 로직이 필요하냐고 코멘트를 남겼었지요. 그 이유는 테스트의 맥락상 절대 예외가 발생하지 않을 상황이기 때문이었어요. 그렇기 때문에 (싫어하실 수도 있지만) 바로 `Optional.get()`으로 꺼내도 되지 않을까 생각했어요. 테스트 코드가 잘못 수정되어 테스트가 실패하더라도 `NoSuchElementException`이 발생할 텐데 원인을 파악하기 어렵지 않겠죠. (물론 프로덕션 코드에서는 절대 그러면 안 됩니다.) 테스트 코드에 지금처럼 로직이 추가로 들어가면 어쨌든 읽고 이해하는 데 리소스가 필요하니까요.

물론 이건 제 의견이라 꼭 반영해달라는 말씀을 드리는 건 절대 아니에요!

tco0427 replied:
코니가 말씀해주신대로 지금처럼 로직이 추가되면 비용이 소모된다는데에도 동의하며, get() 을 활용하면  테스트가 실패하는 경우 `NoSuchElementException`이 발생하여 원인 파악에 도움이 될 것 같습니다!!
하지만 여전히 단순히 get()을 사용하는 것은..조금 마음에 걸리는 것 같아요..(IntelliJ 경고(?)도 그렇고...🥲)
혹시 그 비용을 조금 더 줄일 수 있게 다음과 같이 "부적절한 Line" 이라는 Line을 미리 생성해두고 활용한다면 지금보다 조금 더 괜찮을까요??
```java
    private static final String FAIL_FIND_LINE = "fail";
    private static final Line INAPPROPRIATE_LINE = new Line(FAIL_FIND_LINE, FAIL_FIND_LINE);
```
```java
        final Line findLine = jdbcLineDao.findById(savedLine.getId())
                .orElse(INAPPROPRIATE_LINE);
```

choihz commented on src/test/java/wooteco/subway/domain/StationTest.java:
```diff
@@ -1,16 +1,27 @@
 
+    private static final String EXCESS_MAX_LENGTH_STRING = "-".repeat(256);
+
     @DisplayName("지하철역의 이름이 공백인지를 검사한다.")
```
```suggestion
    @DisplayName("지하철역의 이름은 공백일 수 없다.")
```
또는
```suggestion
    @DisplayName("지하철역의 이름이 공백이면 예외가 발생한다.")
```

이렇게는 어때요?

tco0427 replied:
좋습니다!! 다른 단위 테스트들의 `@DisplayName` 에서도 "~~ 수 없다." 와 같은 형식을 취했으므로 일관성 있게 "지하철역의 이름은 공백일 수 없다."로 수정해보도록 하겠습니다.

choihz commented on src/test/java/wooteco/subway/domain/StationTest.java:
```diff
@@ -1,16 +1,27 @@
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @DisplayName("지하철역은 255보다 크거나 같을 수 없다.")
```
```suggestion
    @DisplayName("지하철역의 이름은 255보다 크거나 같을 수 없다.")
```

그런데 크거나 같을 수 없다면 "같은" 상황도 검증해 봐야 하는 것 아닐까요?

tco0427 replied:
255자 까지는 가능하고 256자부터 불가능하므로 "지하철역의 이름은 255보다 클 수 없다."로 수정하였습니다!😃

choihz commented on src/main/java/wooteco/subway/domain/vo/StationName.java:
```diff
@@ -0,0 +1,48 @@
+        if (name.isBlank()) {
+            throw new IllegalArgumentException("지하철역 이름이 공백일 수 없습니다.");
+        }
+        if (name.length() > 255) {
+            throw new IllegalArgumentException("지하철역 이름은 255자 보다 크거나 같을 수 없습니다.");
+        }
```
같을 수 없는 게 맞나요? 🤔

tco0427 replied:
앗...! 조건문을 변경해놓고, 예외 메시지는 체크하지 못했네요ㅠ.ㅠ 수정하였습니다😅

choihz commented on src/test/java/wooteco/subway/domain/StationTest.java:
```diff
@@ -1,16 +1,27 @@
+    @DisplayName("지하철역은 255보다 크거나 같을 수 없다.")
+    @Test
+    public void stationNameLengthTest() {
+        // given & when & then
```
`// given & when & then` 주석은 불필요해 보이네요 😅

tco0427 replied:
이러한 경우에는 보통 테스트가 하나의 라인으로 끝나는데, given & when & then을 구분해줄 필요가 없으므로 해당 주석은 불필요해보이네요! 수정해보도록 하겠습니다!

choihz commented on src/test/java/wooteco/subway/domain/StationTest.java:
```diff
@@ -1,16 +1,27 @@
+        assertThatThrownBy(() -> new Station(EXCESS_MAX_LENGTH_STRING))
                 .isInstanceOf(IllegalArgumentException.class);
     }
 }
```
<img width="420" alt="스크린샷 2022-05-13 오전 3 03 28" src="https://user-images.githubusercontent.com/18049492/168140020-fa270eb1-9117-4733-8e45-f4e20a29bd2b.png">

이 설정을 켜는 건 어떨까요? 왜 필요한지는 혹시 알고 계신가요?

tco0427 replied:
넵..!! 이전에도 동일한 [리뷰](https://github.com/woowacourse/java-chess/pull/358#discussion_r844562308)를 받았었습니다!

EOF 와 관련된 문제가 발생할 수 있을 것 같습니다..!
당시에 어떤 옵션을 적용해야할지 몰랐었는데, 코니 덕분에 설정 적용해볼 수 있었습니다!! 감사합니다.😃

choihz commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -24,16 +23,16 @@ class LineAcceptanceTest extends AcceptanceTest {
     void createLine() {
         // given
-        final LineRequest params = new LineRequest("신분당선", "bg-red-600");
+        final ExtractableResponse<Response> upStationResponse = makeStationResponse("지하철역");
+        final Long upStationId = extractStationId(upStationResponse);
```
> 1. 지하철역 이름(name)을 통해서 `StationRequest` 를 만들고, 요청을 보낸 이후 `ExtractableResponse` 를 만든다.
> 2. 해당 응답에서 id값을 추출한다.
> 
> 이렇게 2가지 책임을 가지기 때문에 해당 메소드가 하는 일을 `메소드 이름`을 통해서 얻기가 어려웠다고 생각이 듭니다.
> 두 책임은 엄연히 다른 책임이고, `makeStationResponse` 와 `extractStationId` 두 가지 메소드로 분리하여 `메소드 이름`을 통해서 그 의미를 보다 명확히 알 수 있도록 개선해보았습니다!😃

그런데 이 두 메서드는 항상 같이 사용되고 있지 않나요? 이 둘을 분리했을 때의 장점을 느끼셨을까요?
이전에 제가 남긴 코멘트는, 메서드가 하는 본질적인 역할을 이름을 통해 잘 드러내면 좋다는 의미로 이해해 주시면 좋을 것 같아요!

tco0427 replied:
이전에 말씀드린 것과 같이...메소드가 하는 역할을 `이름`을 통해서 잘 표현하지 못한 이유는 두 가지 역할을 하는 메소드를 하나의 이름으로 표현하려고 하였기 때문이라는 생각이 아직까지 들어요..그래서 저는 차라리 모호한 메소드 이름을 부여하는 것보다는 이렇게 구분하여 보다 명확하게 해주는 것이 맞다는 생각이 들었습니다..!
코니가 말씀해주신대로 항상 함께 사용되기 때문에 하나의 메소드로 표현할 수 있겠는데요! 이렇게 된다면 메소드명을 어떻게 해야할지 모르겠어요ㅠㅠ
지금 생각나는 메소드명은 `extractStationIdFromResponse(String name)` 정도인거 같아요..하지만 `fromResponse` 라는 이름에서 인자로 ExtractableResponse가 와야할 것 같아요🥲

choihz commented on src/test/java/wooteco/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,212 @@
+        assertDoesNotThrow(() -> sections.add(section));
+    }
+    
+    @DisplayName("하나의 노선에는 갈래길이 허용되지 않기 때문에 새로운 구간이 추가되기 전에 갈래길이 생기지 않도록 기존 구간을 변경한다.")
```
바로 아래 있는 메서드와 이름이 동일하네요 👀

tco0427 replied:
`어떤 역을 기준으로 하는지` 조금 더 구체적으로 변경해주어 두 단위 테스트가 어떻게 다른지를 분명히 해주었습니다.😄

choihz commented on src/test/java/wooteco/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,212 @@
+    public void forkRodeSameUpStation() {
+        // given
+        List<Section> sectionList = new ArrayList<>();
+        sectionList.add(new Section(1L, 1L, 1L, 3L, 7));
```
```suggestion
        sectionList.add(new Section(1L, 1L, 3L, 7));
```

`Section`의 식별자는 이 맥락에선 불필요해 보이는데 제거해도 되지 않을까요? 비슷한 숫자들이 연달아 있다 보니 각 값의 의미를 이해하는 데 시간이 걸리는 것 같아요.

tco0427 replied:
<img width="507" alt="스크린샷 2022-05-14 오전 9 10 10" src="https://user-images.githubusercontent.com/57028386/168403064-8ab18969-fcd7-46d9-ab97-357ee1af6e23.png">
위와 같이 IDE의 도움을 받다 보니 고려해보지 못했던 부분이네요...😅
말씀해주신 것과 같이 해당 단위 테스트에서 Section의 식별자는 불필요하고, 이를 생략함으로써 코드를 읽는 사람으로 하여금 관심있는 부분만 집중하게 할 수 있을 것 같습니다! 수정해보겠습니다~_~

tco0427 replied:
나머지 부분에서도 Section의 식별자는 불필요하므로 제거하였습니다! 하지만 예외 적으로 `checkDistance()` 와 `sameSection()` 에서는 무엇에 의해서 예외가 발생하는지 분명히 해주고 싶어 Section의 id 값을 그대로 사용하도록 유지시켜 주었습니다!

choihz commented on src/test/java/wooteco/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,212 @@
+    }
+    
+    @DisplayName("하나의 노선에는 갈래길이 허용되지 않기 때문에 새로운 구간이 추가되기 전에 갈래길이 생기지 않도록 기존 구간을 변경한다.")
+    @Test
+    public void forkRodeSameUpStation() {
+        // given
+        List<Section> sectionList = new ArrayList<>();
+        sectionList.add(new Section(1L, 1L, 1L, 3L, 7));
+        final Sections sections = new Sections(sectionList);
+
+        // when
+        final Section section = new Section(2L, 1L, 1L, 2L, 4);
+        sections.add(section);
+
+        // then
+        assertThat(sections.getSections().size()).isEqualTo(2);
+        assertThat(sections.getSections()).extracting("distance")
+                .containsExactlyInAnyOrder(4, 3);
+    }
```
`1 - (7) - 3`에서 `1 - (3) - 2 - (4) - 3`이 되는데, `getSections()`의 리턴 순서는 삽입 순서와 동일하여 거리가 순서대로 4, 3임을 검증하고 있네요.

그런데 `Sections`를 개념적으로 잘 생각해 보면, 본질적으로 순서가 있어야 하는 것 아닐까요? 지금은 `LineService`에서 노선을 리턴할 때만 `getSortedSection()` 메서드를 이용해 정렬을 하고 있는데 적절한 상황일까요?

tco0427 replied:
Sections는 어떤 라인의 구간들을 나타내어 주는 도메인이기 때문에 말씀해주신대로 순서를 정렬한 상태로 유지하는게 맞는 것 같습니다..!! 따라서 Sections를 생성하면서 정렬된 형태를 유지하고, add, delete 시에도 정렬을 해주도록 하였습니다. 또한 테스트에서도 containsExactly(3, 4) 와 같이 수정하여 구간의 크기 뿐 아니라 그 순서도 정확하게 테스트 하도록 수정해보았습니다.😄

choihz commented on src/test/java/wooteco/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,212 @@
+        sections.add(section);
+
+        // then
+        assertThat(sections.getSections().size()).isEqualTo(2);
+        assertThat(sections.getSections()).extracting("distance")
+                .containsExactlyInAnyOrder(4, 3);
```
```suggestion
        assertThat(sections.getSections())
                .hasSize(2)
                .extracting("distance")
                .containsExactlyInAnyOrder(4, 3);
```

이렇게 체이닝도 가능해요.

tco0427 replied:
적용해보았습니다!!🙂

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,211 @@
+                newSection.getUpStationId(), existingSection.getDistance() - newSection.getDistance());
+    }
+
+    private void isPossibleRegistration(Section section) {
+        sections.stream()
+                .filter(s -> s.containsStation(section))
+                .findAny()
+                .orElseThrow(() -> new IllegalSectionException("등록할 구간의 적어도 하나의 역은 등록되어 있어야 합니다."));
+    }
```
이런 이름의 메서드는 보통 단순히 `boolean`을 리턴하죠!

tco0427 replied:
앗...! 맞습니다...ㅠ.ㅠ `validateRegistration`으로 변경하여 "등록가능한지를 검증한다." 라는 의미를 전달하도록 수정하여 반환형에 대한 오해가 없도록 수정해보았습니다!

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    private void checkContainsSameSection(Section newSection) {
+        final boolean isContains = sections.stream()
+                .anyMatch(section -> section.isSameSection(newSection));
```
여기도 is + 동사원형!

tco0427 replied:
이 부분도 저번에 관련 코멘트를 주셨을 때 수정했던 부분인데, 제대로 확인하지 못했던 것 같네요..ㅠ.ㅠ contains 라고 수정해볼께요!! `if (contains) {...}` 라고 하면 해당 변수를 사용하는 if 문에서 "만약 포함하면" 이라고 해석이 될 수 있을 것 같아요!

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,211 @@
+        sections.remove(existingSection);
+    }
+
+    private void checkDistance(Section existingSection, Section newSection) {
+        if (existingSection.getDistance() <= newSection.getDistance()) {
+            throw new IllegalSectionException("등록하려는 구간 길이가 기존 구간의 길이보다 더 길 수 없습니다.");
+        }
+    }
```
예외 메시지가 길이가 같은 상황은 포괄하지 못하고 있네요.

tco0427 replied:
앗,,제가 예외 메시지를 포함한 이러한 메시지들에서 꼼꼼하지 못한 것 같네요..🥲
습관화할 수 있도록 노력하겠습니다!! 감사합니다!🙏

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,211 @@
+
+    private Section getPreviousSection(Long stationId) {
+        return sections.stream()
+                .filter(section -> section.getDownStationId().equals(stationId))
```
값을 가져와서 비교하지 않고 물어볼 수 있을 것 같아요.

tco0427 replied:
`isSameDownStation()`, `isSameUpStation()` 이라는 메소드를 `section`에 만들어 값을 가져오는 것이 아니라 메시지를 보내도록 수정해보았습니다!!

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,211 @@
+                .orElseThrow(() -> new IllegalSectionException("등록할 구간의 적어도 하나의 역은 등록되어 있어야 합니다."));
+    }
+
+    public List<Section> delete(Long stationId) {
+        validateThatHasMinimumSection();
+
+        final Section previousSection = getPreviousSection(stationId);
+        final Section laterSection = getLaterSection(stationId);
+
+        findAndRemoveFirstSection(stationId, previousSection);
+        findAndRemoveLastSection(stationId, laterSection);
+        removeMiddleSection(previousSection, laterSection);
+
+        return sections;
+    }
```
이 메서드 로직을 보면, 상행 종점을 지우는 경우 `laterSection`만 찾을 수 있고, 하행 종점을 지우는 경우 `previousSection`만 찾을 수 있고, 중간에 있는 역을 지우는 경우 둘 다 찾을 수 있을 것 같은데요. 그러면 이미 `previousSection`과 `laterSection`의 존재 여부만으로 종점 여부를 알 수 있고 삭제할 구간도 바로 찾은 셈이니 이에 따라 로직을 전개하면 되지 않나요? `findAndRemoveFirstSection()` 등의 로직을 보면 내부에서 또 삭제할 구간을 다시 찾고 있는데 이렇게 할 필요가 있는 건가요?

제가 로직을 잘못 이해한 부분이 있다면 코멘트 부탁드려요!

tco0427 replied:
`getPreviousSection()` 과 `getLaterSection()` 에서 지우려는 역이 만약 종점의 경우 둘 중 하나가 null이 되게 되는데요,
만약 `getPreviousSection()`이 null인 경우(상행 종점 제거) 즉, `1 - 2 - 3 - 4` 에서 1을 지우려고 하는 경우, getPreviousSection이 null이기 때문에 `첫번째 구간`을 찾는 `getFirstSection()`이 필요하게 됩니다.

그런데 앞선 코멘트를 반영하면서 Sections가 항상 정렬된 상태를 유지하기 때문에 삭제할 구간을 찾을 필요 없이 상행 종점 제거의 경우 `index: 0`, 하행 종점 제거의 경우 `index: sections.size()-1` 을 이용할 수 있을 것 같습니다!

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,211 @@
+        return sections;
+    }
+
+    private void validateThatHasMinimumSection() {
+        if (sections.size() <= MINIMUM_SECTION_COUNT) {
+            throw new IllegalSectionException("노선이 구간을 하나는 가져야하므로 구간을 제거할 수 없습니다.");
+        }
+    }
```
> 고민을 해보았는데요...`validateMinimumSectionCount`, `validateThatHasMinimumSection` 를 후보(?)로 했습니다...의미만 보면 `validateThatHasMinimumSection` 이 나은 것 같은데 메소드명이 너무 길어도 안좋지 않나??하는 생각이 들었어요. 코니는 어떻게 생각하시나요??
> (hasMinimumSection도 괜찮은 것 같다는 생각도 듭니다.🙂)

지적만 하는 게 역할인 제게 어려운 퀘스트를 주시는군요 🤣
저라면… `validateDeletableSize()`…? 참고만 해 주세요 😇

tco0427 replied:
오!!! validateDeletableSize() 너무 좋은거 같아요!! 해당 네이밍으로 수정해볼게요!! 함께 고민해주셔서 감사합니다~_~

choihz reviewed (changes requested):
안녕하세요, 디우!

수정해 주신 부분을 확인하고 몇 가지 코멘트를 더 남겨 보았습니다. 궁금한 점 또는 의견이 있으실 땐, 아시죠? 😌

tco0427 committed:
refactor: DisplayName 수정

tco0427 committed:
refactor: 예외 메시지 수정

tco0427 committed:
refactor: 불필요한 given when then 주석 제거

tco0427 committed:
refactor: 파일의 끝에 개행 추가

tco0427 committed:
refactor: DisplayName 구체적으로 수정

tco0427 committed:
refactor: Section 식별자 제거

Section의 식별자가 불필요한 곳에서 제거해주었습니다.

tco0427 committed:
refactor: INAPPROPRIATE_LINE 추가

Optional이 비어있을 미리 생성해둔 상수(static final)에 해당하는 INAPPROPRIATE_LINE을 반환

tco0427 committed:
refactor: Sections 정렬된 상태 유지

Sections가 생성되는 시점 그리고 데이터 변경이 일어나는 add, delete 순간에도 정렬을 수행하도록 수정하였습니다.

tco0427 committed:
refactor: size검사와 내용 검사 하나로 체이닝

tco0427 committed:
refactor: 부적절한 isXXX 수정 및 제거

tco0427 committed:
refactor: 예외 메시지 변경

tco0427 committed:
refactor: 같은 Station 비교 구현

값을 직접 가져와서 비교하는게 아니라 Section에 같은 Station인지 비교하는 메소드를 두어, 물어보도록 수정하였습니다.

tco0427 committed:
refactor: validateThatHasMinimumSection -> validateDeletableSize 네이밍 변경

tco0427 committed:
refactor: 구간 삭제 로직 수정

Sections가 항상 정렬된 상태를 유지하므로 종점 삭제시 index 0과 size()-1 을 이용함으로써 구간을 찾는 로직을 제거해주었습니다.

choihz commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -27,6 +24,9 @@ private void validateArgument(String name, String color) {
         if (name.isBlank() || color.isBlank()) {
             throw new IllegalArgumentException("노선의 이름 혹은 색이 공백일 수 없습니다.");
         }
+        if (name.length() >= 255 || color.length() >= 20) {
+            throw new IllegalArgumentException("노선의 이름 혹은 색이 너무 깁니다.");
+        }
```
여기도 메시지에 좀 더 신경 써 주세요~

tco0427 replied:
```java
        if (name.length() > 255 || color.length() > 20) {
            throw new IllegalArgumentException("노선의 이름이 255자 보다 크거나, 색이 20자 보다 큽니다.");
        }
```
위와 같이 수정해보았습니다!!

choihz commented on src/main/java/wooteco/subway/domain/Section.java:
```diff
@@ -0,0 +1,84 @@
+
+public class Section {
+
+    private final Long id;
+    private final Long lineId;
+    private final Long upStationId;
+    private final Long downStationId;
+    private final int distance;
```
`Section`의 필드에는 모두 `final`이 붙어 있는데 `Line`, `Station`에는 누락되어 있네요 😓

tco0427 replied:
이런부분도 체크하지 못했었다니..부끄럽네요ㅜ.ㅜ😅
수정해보도록 할게요!!

choihz commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -2,43 +2,64 @@
+import wooteco.subway.dto.StationResponse;
 import wooteco.subway.exception.NotExistException;
 
 @Service
+@Transactional
 public class LineService {
```
`LineService`, `StationService`는 클래스 레벨의 `@Transactional`을 가지고 있고 `SectionService`는 그렇지 않은데요, 여기도 일관성이 지켜지면 좋을 것 같아요.

tco0427 replied:
동의합니다!! 통일해볼게요~_~

choihz commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,192 @@
+
+    private List<Section> sections;
+
+    public Sections(List<Section> sections) {
+        this.sections = sections;
+        this.sections = sortSections(sections);
+    }
```
이렇게 두 번 할당할 필요가 있나요? 👀

tco0427 replied:
Sections의 정렬하는 부분에서, 특히 첫번째 section인지를 검사하는 부분에서 인스턴스 변수인 sections와 비교를 하기 때문에..먼저 할당한 이후 sortSections를 불러야 해서 이렇게 두 번 할당하게 되어 있습니다..!

```java
    private boolean isFirst(Long upStationId) {
        return sections.stream()
                .noneMatch(section -> upStationId.equals(section.getDownStationId()));
    }
```

저도 리팩토링 하면서 이부분이 찜찜하긴 했는데, 한 번에 전부 다 바꾸려고 하면 제대로 개선하지 못할 것 같아서 우선 이렇게 수정해보았는데요! 이 부분도 한 번만 할당해도 되게끔 리팩토링 해보도록 하겠습니다!😁

choihz commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -53,5 +74,44 @@ public void deleteById(Long id) {
+
+        List<StationResponse> stations = new ArrayList<>();
+
+        for (final Section section : sortedSections) {
+            final Station findStation = getStationById(section.getUpStationId());
+            stations.add(new StationResponse(findStation));
+        }
```
`for`문 안에서 DB에 접근하여 역의 정보를 하나하나 가져오고 있는데요, 한번에 가져올 수 있는 방법은 없는지 고민해 보면 좋을 것 같아요.

tco0427 replied:
조금 오래 고민해보았는데요! section으로 부터 StationId들을 중복없이 TreeSet으로 받고, 이 id들을 IN 절을 사용해서 조회해볼 수 있을 것 같아요!!😃

choihz commented on src/test/java/wooteco/subway/dao/jdbc/JdbcLineDaoTest.java:
```diff
@@ -61,10 +64,11 @@ void findAll() {
 
         // when
-        final Line findLine = lineDao.findById(savedLine.getId());
+        final Line findLine = jdbcLineDao.findById(savedLine.getId())
+                .orElse(INAPPROPRIATE_LINE);
```
> 코니가 말씀해주신대로 지금처럼 로직이 추가되면 비용이 소모된다는데에도 동의하며, get() 을 활용하면  테스트가 실패하는 경우 `NoSuchElementException`이 발생하여 원인 파악에 도움이 될 것 같습니다!!
> 하지만 여전히 단순히 get()을 사용하는 것은..조금 마음에 걸리는 것 같아요..(IntelliJ 경고(?)도 그렇고...🥲)
> 혹시 그 비용을 조금 더 줄일 수 있게 다음과 같이 "부적절한 Line" 이라는 Line을 미리 생성해두고 활용한다면 지금보다 조금 더 괜찮을까요??

네 괜찮은 것 같아요 👍

choihz commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -24,16 +23,16 @@ class LineAcceptanceTest extends AcceptanceTest {
     void createLine() {
         // given
-        final LineRequest params = new LineRequest("신분당선", "bg-red-600");
+        final ExtractableResponse<Response> upStationResponse = makeStationResponse("지하철역");
+        final Long upStationId = extractStationId(upStationResponse);
```
> 이전에 말씀드린 것과 같이...메소드가 하는 역할을 `이름`을 통해서 잘 표현하지 못한 이유는 두 가지 역할을 하는 메소드를 하나의 이름으로 표현하려고 하였기 때문이라는 생각이 아직까지 들어요..그래서 저는 차라리 모호한 메소드 이름을 부여하는 것보다는 이렇게 구분하여 보다 명확하게 해주는 것이 맞다는 생각이 들었습니다..!
> 코니가 말씀해주신대로 항상 함께 사용되기 때문에 하나의 메소드로 표현할 수 있겠는데요! 이렇게 된다면 메소드명을 어떻게 해야할지 모르겠어요ㅠㅠ
> 지금 생각나는 메소드명은 `extractStationIdFromResponse(String name)` 정도인거 같아요..하지만 `fromResponse` 라는 이름에서 인자로 ExtractableResponse가 와야할 것 같아요🥲

`extract`라는 맥락이 중요한지, 이 메서드를 가져다 사용하는 입장에서 꼭 알아야 하는 정보는 무엇인지 고민해 보면 좋을 것 같아요~

tco0427 replied:
사용하는 입장에서 보면 `역 이름을 준다.`(input), `해당 역의 id 값을 갖는다.`(output) 이라는 정보가 중요할 것 같아요!
그러한 맥락에서 이전의 두 메소드를 합치고, 그 메소드의 이름으로 `extractIdByStationName()`는 어떨까하는 생각이 듭니다!

choihz reviewed (approved):
안녕하세요, 디우!

피드백을 반영해 주신 내용 모두 확인하고, 몇 가지 코멘트를 남겨 두었습니다 👍 이제 지하철 노선도 미션은 마무리해도 좋을 것 같아요. 그동안 정말 고생 많으셨고, 함께하는 과정이 즐거웠어요. 그럼, 다음에 다시 만나요~ 👋
